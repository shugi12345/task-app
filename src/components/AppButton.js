import React from 'react';
import { TouchableOpacity, Text, StyleSheet } from 'react-native';

export default function AppButton({ title, onPress, style }) {
  return (
    <TouchableOpacity style={[styles.button, style]} onPress={onPress}>
      <Text style={styles.text}>{title}</Text>
    </TouchableOpacity>
  );
}

const styles = StyleSheet.create({
  button: {
    backgroundColor: '#bb86fc',
    paddingVertical: 10,
    paddingHorizontal: 12,
    borderRadius: 8,
    alignItems: 'center',
    marginHorizontal: 4,
  },
  text: {
    color: '#fff',
    fontSize: 16,
    fontWeight: 'bold',
  },
});
