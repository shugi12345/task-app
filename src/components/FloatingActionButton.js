import React from 'react';
import { TouchableOpacity, Text, StyleSheet } from 'react-native';

export default function FloatingActionButton({ onPress }) {
  return (
    <TouchableOpacity style={styles.fab} onPress={onPress}>
      <Text style={styles.fabText}>+</Text>
    </TouchableOpacity>
  );
}

const styles = StyleSheet.create({
  fab: {
    position: 'absolute',
    right: 24,
    bottom: 24,
    backgroundColor: '#bb86fc',
    width: 56,
    height: 56,
    borderRadius: 8,
    alignItems: 'center',
    justifyContent: 'center',
    elevation: 4,
  },
  fabText: {
    color: '#fff',
    fontSize: 32,
    lineHeight: 32,
  },
});
