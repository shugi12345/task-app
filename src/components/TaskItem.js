import React from 'react';
import {
  View,
  Text,
  StyleSheet,
  Animated,
  TouchableOpacity,
} from 'react-native';
import Checkbox from 'expo-checkbox';

function colorFor(level) {
  const ratio = (level - 1) / 4;
  const r = Math.round(Math.min(255, 510 * ratio));
  const g = Math.round(Math.min(255, 510 * (1 - ratio)));
  return `rgb(${r}, ${g}, 0)`;
}

function computeUrgency(task) {
  if (!task.dueDate) return task.urgency;
  const due = new Date(task.dueDate);
  const created = new Date(task.created);
  const now = new Date();
  const total = due - created;
  const remaining = due - now;
  if (total <= 0) return task.urgency;
  const progress = 1 - remaining / total;
  return Math.min(5, task.urgency + Math.floor(progress * (5 - task.urgency)));
}

export default function TaskItem({ task, onComplete, onPress }) {
  const [checked, setChecked] = React.useState(false);
  const anim = React.useRef(new Animated.Value(0)).current;
  const urgency = computeUrgency(task);

  React.useEffect(() => {
    if (checked) {
      Animated.timing(anim, {
        toValue: 1,
        duration: 300,
        useNativeDriver: true,
      }).start(() => onComplete());
    }
  }, [checked]);

  return (
    <TouchableOpacity activeOpacity={0.7} onPress={onPress}>
      <Animated.View
        style={[
          styles.row,
          {
            opacity: anim.interpolate({ inputRange: [0, 1], outputRange: [1, 0] }),
            transform: [
              {
                translateX: anim.interpolate({ inputRange: [0, 1], outputRange: [0, -100] }),
              },
            ],
          },
        ]}
      >
        <Checkbox style={styles.checkbox} value={checked} onValueChange={setChecked} />
        <View style={styles.info}>
          <Text style={styles.title}>{task.title}</Text>
          <Text style={styles.duration}>{task.duration}</Text>
          {task.dueDate && (
            <Text style={styles.dueDate}>Due {task.dueDate.slice(0, 10)}</Text>
          )}
        </View>
        <View style={styles.urgency}>
          {[1, 2, 3, 4, 5].map((l) => (
            <View
              key={l}
              style={[
                styles.dot,
                { backgroundColor: l <= urgency ? colorFor(urgency) : '#333' },
              ]}
            />
          ))}
        </View>
      </Animated.View>
    </TouchableOpacity>
  );
}

const styles = StyleSheet.create({
  row: {
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: '#1e1e1e',
    paddingVertical: 12,
    paddingRight: 8,
    paddingLeft: 16,
    borderRadius: 8,
    marginVertical: 6,
  },
  info: {
    flex: 1,
    marginLeft: 8,
  },
  title: {
    fontSize: 16,
    fontWeight: 'bold',
    color: '#fff',
  },
  duration: {
    fontSize: 12,
    color: '#aaa',
  },
  dueDate: {
    fontSize: 12,
    color: '#aaa',
  },
  urgency: {
    marginLeft: 8,
    flexDirection: 'row',
  },
  dot: {
    width: 12,
    height: 12,
    borderRadius: 6,
    marginHorizontal: 2,
  },
  checkbox: {
    width: 20,
    height: 20,
  },
});
