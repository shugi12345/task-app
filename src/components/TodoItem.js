import React from 'react';
import { Animated, StyleSheet, TextInput, View, TouchableOpacity } from 'react-native';
import Checkbox from 'expo-checkbox';

export default function TodoItem({ item, onToggle }) {
  const [checked, setChecked] = React.useState(false);
  const slideAnim = React.useRef(new Animated.Value(item.animateIn ? -1 : 0)).current;

  React.useEffect(() => {
    if (item.animateIn) {
      Animated.timing(slideAnim, {
        toValue: 0,
        duration: 300,
        useNativeDriver: true,
      }).start();
    }
  }, []);

  React.useEffect(() => {
    if (checked) {
      Animated.timing(slideAnim, {
        toValue: 1,
        duration: 300,
        useNativeDriver: true,
      }).start(onToggle);
    }
  }, [checked]);

  return (
    <Animated.View
      style={[
        styles.item,
        {
          opacity: slideAnim.interpolate({ inputRange: [-1, 0, 1], outputRange: [0, 1, 0] }),
          transform: [
            {
              translateX: slideAnim.interpolate({
                inputRange: [-1, 0, 1],
                outputRange: [-100, 0, -100],
              }),
            },
          ],
        },
      ]}
    >
      <TouchableOpacity onPress={() => setChecked(!checked)} style={styles.checkboxWrapper}>
        <Checkbox
          style={styles.checkbox}
          value={checked}
          onValueChange={setChecked}
        />
      </TouchableOpacity>
      <View style={styles.itemTextContainer}>
        <TextInput value={item.text} editable={false} style={styles.itemText} />
      </View>
    </Animated.View>
  );
}

const styles = StyleSheet.create({
  item: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: 12,
    paddingRight: 8,
    paddingLeft: 16,
    backgroundColor: '#1e1e1e',
    marginVertical: 6,
    borderRadius: 8,
  },
  itemTextContainer: {
    flex: 1,
    marginLeft: 16,
  },
  itemText: {
    fontSize: 16,
    color: '#fff',
  },
  checkboxWrapper: {
    padding: 8,
  },
  checkbox: {
    width: 20,
    height: 20,
  },
});
