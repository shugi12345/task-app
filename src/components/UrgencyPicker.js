import React from 'react';
import { View, Pressable, StyleSheet } from 'react-native';

function colorFor(level) {
  const ratio = (level - 1) / 4; // 0 -> green, 1 -> red
  const r = Math.round(Math.min(255, 510 * ratio));
  const g = Math.round(Math.min(255, 510 * (1 - ratio)));
  return `rgb(${r}, ${g}, 0)`;
}

export default function UrgencyPicker({ value, onChange }) {
  return (
    <View style={styles.row}>
      {[1,2,3,4,5].map(l => (
        <Pressable
          key={l}
          onPress={() => onChange(l)}
          style={[
            styles.circle,
            { backgroundColor: l <= value ? colorFor(value) : '#555' },
          ]}
        />
      ))}
    </View>
  );
}

const styles = StyleSheet.create({
  row: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginVertical: 8,
  },
  circle: {
    width: 20,
    height: 20,
    borderRadius: 10,
    marginHorizontal: 2,
  },
});
