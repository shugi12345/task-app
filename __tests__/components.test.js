import React from 'react';
import renderer, { act } from 'react-test-renderer';
import AppButton from '../src/components/AppButton';
import FloatingActionButton from '../src/components/FloatingActionButton';
import TaskItem from '../src/components/TaskItem';
import TodoItem from '../src/components/TodoItem';
import UrgencyPicker from '../src/components/UrgencyPicker';

jest.mock('expo-checkbox', () => 'Checkbox');

const sampleTask = {
  title: 'Test',
  duration: '1h',
  urgency: 3,
  created: '2023-01-01T00:00:00Z',
  dueDate: '2023-01-03T00:00:00Z',
};

describe('Component snapshots', () => {
  test('AppButton', () => {
    let tree;
    act(() => {
      tree = renderer.create(<AppButton title="OK" onPress={() => {}} />);
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  test('FloatingActionButton', () => {
    let tree;
    act(() => {
      tree = renderer.create(<FloatingActionButton onPress={() => {}} />);
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  test('TaskItem', () => {
    let tree;
    act(() => {
      tree = renderer.create(
        <TaskItem task={sampleTask} onComplete={() => {}} onPress={() => {}} />
      );
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  test('TodoItem', () => {
    let tree;
    act(() => {
      tree = renderer.create(
        <TodoItem item={{ id: '1', text: 'Do', animateIn: false }} onToggle={() => {}} />
      );
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  test('UrgencyPicker', () => {
    let tree;
    act(() => {
      tree = renderer.create(<UrgencyPicker value={3} onChange={() => {}} />);
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
