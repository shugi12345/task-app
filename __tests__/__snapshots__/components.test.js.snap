// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component snapshots AppButton 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#bb86fc",
      "borderRadius": 8,
      "marginHorizontal": 4,
      "opacity": 1,
      "paddingHorizontal": 12,
      "paddingVertical": 10,
    }
  }
>
  <Text
    style={
      [
        {
          "color": "#fff",
          "fontSize": 16,
          "fontWeight": "bold",
        },
        undefined,
      ]
    }
  >
    OK
  </Text>
</View>
`;

exports[`Component snapshots FloatingActionButton 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#bb86fc",
      "borderRadius": 8,
      "bottom": 32,
      "elevation": 4,
      "height": 56,
      "justifyContent": "center",
      "opacity": 1,
      "position": "absolute",
      "right": 32,
      "width": 56,
    }
  }
>
  <Text
    style={
      {
        "color": "#fff",
        "fontSize": 32,
        "lineHeight": 32,
      }
    }
  >
    +
  </Text>
</View>
`;

exports[`Component snapshots TaskItem 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
>
  <View
    collapsable={false}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#1e1e1e",
        "borderRadius": 8,
        "flexDirection": "row",
        "marginVertical": 6,
        "opacity": 1,
        "paddingLeft": 16,
        "paddingRight": 8,
        "paddingVertical": 12,
        "transform": [
          {
            "translateX": 0,
          },
        ],
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
          "padding": 8,
        }
      }
    >
      <Checkbox
        onValueChange={[Function]}
        style={
          {
            "height": 20,
            "width": 20,
          }
        }
        value={false}
      />
    </View>
    <View
      style={
        {
          "flex": 1,
          "marginLeft": 16,
        }
      }
    >
      <Text
        style={
          {
            "color": "#fff",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        Test
      </Text>
      <Text
        style={
          {
            "color": "#aaa",
            "fontSize": 12,
          }
        }
      >
        1h
      </Text>
      <Text
        style={
          {
            "color": "#aaa",
            "fontSize": 12,
          }
        }
      >
        Due 
        03/01/2023
      </Text>
    </View>
    <View
      style={
        {
          "flexDirection": "row",
          "marginLeft": 8,
        }
      }
    >
      <View
        style={
          [
            {
              "borderRadius": 6,
              "height": 12,
              "marginHorizontal": 2,
              "width": 12,
            },
            {
              "backgroundColor": "rgb(255, 0, 0)",
            },
          ]
        }
      />
      <View
        style={
          [
            {
              "borderRadius": 6,
              "height": 12,
              "marginHorizontal": 2,
              "width": 12,
            },
            {
              "backgroundColor": "rgb(255, 0, 0)",
            },
          ]
        }
      />
      <View
        style={
          [
            {
              "borderRadius": 6,
              "height": 12,
              "marginHorizontal": 2,
              "width": 12,
            },
            {
              "backgroundColor": "rgb(255, 0, 0)",
            },
          ]
        }
      />
      <View
        style={
          [
            {
              "borderRadius": 6,
              "height": 12,
              "marginHorizontal": 2,
              "width": 12,
            },
            {
              "backgroundColor": "rgb(255, 0, 0)",
            },
          ]
        }
      />
      <View
        style={
          [
            {
              "borderRadius": 6,
              "height": 12,
              "marginHorizontal": 2,
              "width": 12,
            },
            {
              "backgroundColor": "rgb(255, 0, 0)",
            },
          ]
        }
      />
    </View>
  </View>
</View>
`;

exports[`Component snapshots TodoItem 1`] = `
<View
  collapsable={false}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#1e1e1e",
      "borderRadius": 8,
      "flexDirection": "row",
      "marginVertical": 6,
      "opacity": 1,
      "paddingLeft": 16,
      "paddingRight": 8,
      "paddingVertical": 12,
      "transform": [
        {
          "translateX": 0,
        },
      ],
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
        "padding": 8,
      }
    }
  >
    <Checkbox
      onValueChange={[Function]}
      style={
        {
          "height": 20,
          "width": 20,
        }
      }
      value={false}
    />
  </View>
  <View
    style={
      {
        "flex": 1,
        "marginLeft": 16,
      }
    }
  >
    <TextInput
      editable={false}
      style={
        {
          "color": "#fff",
          "fontSize": 16,
        }
      }
      value="Do"
    />
  </View>
</View>
`;

exports[`Component snapshots UrgencyPicker 1`] = `
<View
  style={
    {
      "flexDirection": "row",
      "justifyContent": "space-between",
      "marginVertical": 8,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "borderRadius": 10,
          "height": 20,
          "marginHorizontal": 2,
          "width": 20,
        },
        {
          "backgroundColor": "rgb(255, 255, 0)",
        },
      ]
    }
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "borderRadius": 10,
          "height": 20,
          "marginHorizontal": 2,
          "width": 20,
        },
        {
          "backgroundColor": "rgb(255, 255, 0)",
        },
      ]
    }
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "borderRadius": 10,
          "height": 20,
          "marginHorizontal": 2,
          "width": 20,
        },
        {
          "backgroundColor": "rgb(255, 255, 0)",
        },
      ]
    }
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "borderRadius": 10,
          "height": 20,
          "marginHorizontal": 2,
          "width": 20,
        },
        {
          "backgroundColor": "#555",
        },
      ]
    }
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "borderRadius": 10,
          "height": 20,
          "marginHorizontal": 2,
          "width": 20,
        },
        {
          "backgroundColor": "#555",
        },
      ]
    }
  />
</View>
`;
